src/Hpack/Config.hs

```diff
@@ -803,39 +804,40 @@ inferModules ::
   -> String
   -> (a -> [String])
   -> (b -> [String])
-  -> ([String] -> [String] -> a -> b)
+  -> (Maybe (List String) -> [String] -> [String] -> a -> b)
   -> ([String] -> a -> b)
   -> Section a
   -> IO (Section b)
 inferModules dir packageName_ getMentionedModules getInferredModules fromData fromConditionals = traverseSectionAndConditionals
   (fromConfigSection fromData [pathsModuleFromPackageName packageName_])
-  (fromConfigSection (\ [] -> fromConditionals) [])
+  (fromConfigSection (\ _ [] -> fromConditionals) [])
   []
   where
     fromConfigSection fromConfig pathsModule_ outerModules sect@Section{sectionData = conf} = do
       modules <- listModules dir sect
+      inferSignatures <- getSignatures dir
```

Comment:
In addition, this looks for signatures next to `package.yaml` (in contrast to `source-dirs`).  My naive assumption would be that signatures have to be inside `source-dirs` too.  But then again, I'm not sure, as documentation is missing and the examples I saw don't specify `hs-source-dirs`.

@ezyang can you clarify?